### PR TITLE
Integrating hardhat-ledger support

### LINKED ISSUE #720: Provide clean UI when using ledger
With the enhancements to `hardhat-ledger` coming from https://github.com/NomicFoundation/hardhat/pull/4946, we can now control the cli UI entirely from Ignition.

## TODO

- [ ] Add `hardhat_setLedgerOutputEnabled` provider rpc call in Ignition deploy to disable ledger cli UI changes
- [ ] Add listeners to the provider that update the UI state
- [ ] Display a waiting for ledger confirmation message at the bottom of the current batch, and update it with successful or failed confirmation
- [ ] Integration test if possible (maybe fake the provider?)

## Potential UI

```shell
Batch #2
  Executed RESOLVER#PublicResolver
  Executed RESOLVER#set_subnode_owner_for_resolver
  Executed REVERSEREGISTRAR#set_subnode_owner_reverse

  Ledger: Connecting wallet
```

```shell
Batch #2
  Executed RESOLVER#PublicResolver
  Executed RESOLVER#set_subnode_owner_for_resolver
  Executed REVERSEREGISTRAR#set_subnode_owner_reverse

  Ledger: Waiting for confirmation
```

```shell
Batch #2
  Executed RESOLVER#PublicResolver
  Executed RESOLVER#set_subnode_owner_for_resolver
  Executed REVERSEREGISTRAR#set_subnode_owner_reverse

  Ledger: Transaction signed
```

### PR DESCRIPTION
**Pull request by zoeyTM**
fixes #720 

Waiting for confirmation from ledger device
Screen Shot 2024-04-09 at 12 18 17 AM

Ledger device confirmed and transaction was sent to network (waiting for block confirmations before continuing)
Screen Shot 2024-04-09 at 12 18 30 AM

Previous ledger messages are removed when execution continues (and are replaced by current ledger messages when needed)
Screen Shot 2024-04-09 at 12 19 21 AM

All ledger messages removed for final status display
Screen Shot 2024-04-09 at 12 20 36 AM

### CODE REVIEW

File: packages/hardhat-plugin/src/ui/pretty-event-handler.ts [hunk 1]
```diff
@@ -309,6 +312,77 @@ export class PrettyEventHandler implements ExecutionEventListener {
...
+      ...this.state,
+      ledger: true,
+      ledgerMessage: "Connecting wallet",
+    };
+
+    this._redisplayCurrentBatch();
+
+    this.state = {
+      ...this.state,
+      ledgerMessageIsDisplayed: true,
```
Comment (kanej) [inline packages/hardhat-plugin/src/ui/pretty-event-handler.ts:326]: Can you walk me through the need for the split here and have two state updates?
Why can't `ledgerMessageIsDisplayed: true` go in the first state update?

File: packages/hardhat-plugin/test/ui/helpers/calculate-deploying-module-panel-display.ts [hunk 2]
```diff
@@ -21,6 +21,9 @@ describe("ui - calculate starting message display", () => {
     maxFeeBumps: 0,
     gasBumps: {},
     strategy: "basic",
+    ledger: false,
+    ledgerMessage: "",
+    ledgerMessageIsDisplayed: false,
```
Comment (kanej) [inline packages/hardhat-plugin/test/ui/helpers/calculate-deploying-module-panel-display.ts:26]: Is the `ledger` property needed in addition to `ledgerMessageIsDisplayed`?
What drives the difference between the `ledger: true` state and the `ledgerMessageIsDisplayed: false` state?

File: packages/hardhat-plugin/src/index.ts [hunk 3]
```diff
@@ -273,6 +314,31 @@ ignitionScope
...
         });
 
+        try {
+          await hre.network.provider.send("hardhat_setLedgerOutputEnabled", [
+            true,
+          ]);
+
+          hre.network.provider.off("connection_start", () =>
+            executionEventListener.ledgerConnectionStart()
+          );
```
Comment (kanej) [inline packages/hardhat-plugin/src/index.ts:324]: The `.off` here won't work as anonymous function is not the same instance as was passed into `.on`.

I suspect that the anonymous functions should be replaced with named local functions, that way both `.on` and `.off` can be passed the same instances.

Comment (zoeyTM) [reply packages/hardhat-plugin/src/index.ts:324]: would this work?

```ts
hre.network.provider.off("connection_start", executionEventListener.ledgerConnectionStart);
```

Comment (kanej) [reply packages/hardhat-plugin/src/index.ts:324]: Yes

**Pull request merged**